File: skills/clean-ddd-hexagonal/examples/typescript/domain/events.ts
// domain/order/events.ts
import { DomainEvent } from './domain_event';
import { OrderId } from './order_id';
import { Money } from './money';

/**
 * Event raised when a new Order is created.
 *
 * This event signals that a customer has initiated a new order. It may
 * trigger downstream processes such as inventory reservation, customer
 * notifications, or analytics tracking.
 */
export class OrderCreated extends DomainEvent {
  /** Event type identifier for routing and serialization. */
  readonly eventType = 'order.created';

  /**
   * Create a new OrderCreated event.
   *
   * @param orderId - The identifier of the newly created order
   * @param customerId - The identifier of the customer who placed the order
   */
  constructor(
    readonly orderId: OrderId,
    readonly customerId: CustomerId,
  ) {
    super();
  }

  /**
   * Convert the event to a serializable payload.
   * @returns Plain object with event data
   */
  toPayload(): Record<string, unknown> {
    return {
      orderId: this.orderId.value,
      customerId: this.customerId.value,
    };
  }
}

/**
 * Event raised when an Order is confirmed for fulfillment.
 *
 * This event indicates the order has been finalized and payment has been
 * processed. It may trigger shipping preparation, inventory deduction,
 * and confirmation notifications.
 */
export class OrderConfirmed extends DomainEvent {
  /** Event type identifier for routing and serialization. */
  readonly eventType = 'order.confirmed';

  /**
   * Create a new OrderConfirmed event.
   *
   * @param orderId - The identifier of the confirmed order
   * @param total - The final monetary total for the order
   */
  constructor(
    readonly orderId: OrderId,
    readonly total: Money,
  ) {
    super();
  }

  /**
   * Convert the event to a serializable payload.
   * @returns Plain object with event data
   */
  toPayload(): Record<string, unknown> {
    return {
      orderId: this.orderId.value,
      total: {
        amount: this.total.amount,
        currency: this.total.currency,
      },
    };
  }
}

/**
 * Event raised when an Order has been shipped.
 *
 * This event indicates the order has been dispatched to the carrier.
 * It may trigger customer notifications with tracking information and
 * update the order status in any read models.
 */
export class OrderShipped extends DomainEvent {
  /** Event type identifier for routing and serialization. */
  readonly eventType = 'order.shipped';

  /**
   * Create a new OrderShipped event.
   *
   * @param orderId - The identifier of the shipped order
   * @param trackingNumber - The carrier's tracking number
   */
  constructor(
    readonly orderId: OrderId,
    readonly trackingNumber: TrackingNumber,
  ) {
    super();
  }

  /**
   * Convert the event to a serializable payload.
   * @returns Plain object with event data
   */
  toPayload(): Record<string, unknown> {
    return {
      orderId: this.orderId.value,
      trackingNumber: this.trackingNumber.value,
    };
  }
}

File: skills/clean-ddd-hexagonal/examples/python/domain/events.py
# src/domain/order/events.py
"""Domain events for the Order aggregate.

Domain events represent significant occurrences within the domain that domain
experts care about. They enable loose coupling between aggregates and support
event-driven architectures.
"""
from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional
from .value_objects import OrderId, CustomerId, Money


@dataclass
class DomainEvent:
    """Base class for all domain events.

    Domain events capture the fact that something significant happened in the
    domain. They are named in past tense (e.g., OrderCreated, PaymentReceived)
    and are immutable records of state changes.

    Attributes:
        occurred_at: Timestamp when the event was created.
    """

    occurred_at: Optional[datetime] = field(default=None)

    def __post_init__(self) -> None:
        """Set the occurred_at timestamp if not provided."""
        if self.occurred_at is None:
            self.occurred_at = datetime.utcnow()


@dataclass
class OrderCreated(DomainEvent):
    """Event raised when a new Order is created.

    This event signals that a customer has initiated a new order, which may
    trigger downstream processes like inventory reservation or customer
    notifications.

    Attributes:
        order_id: The identifier of the newly created order.
        customer_id: The identifier of the customer who placed the order.
    """

    order_id: OrderId = None
    customer_id: CustomerId = None


@dataclass
class OrderConfirmed(DomainEvent):
    """Event raised when an Order is confirmed for fulfillment.

    This event indicates the order has been finalized and payment has been
    processed. It may trigger shipping preparation, inventory deduction,
    and confirmation notifications.

    Attributes:
        order_id: The identifier of the confirmed order.
        total: The final monetary total for the order.
    """

    order_id: OrderId = None
    total: Money = None
